### PR TITLE
arrayfire: 3.7.3 -> 3.9.0

### PR DESCRIPTION
**Pull request by twesterhout**
## Description of changes

To update ArrayFire to the latest release, significant refactoring was necessary.

The derivation contains a lot of `fetchFromGitHub` calls to download external dependencies. Most of these are ArrayFire-specific, and hence it makes little sense to package them separately. The remaining ones, such as "cub" and "spdlog" are rather outdated and hence we again can't rely on the versions in Nixpkgs.

ArrayFire can be built with different backends:
  - CPU: enabled by default and relies on BLAS/LAPACK;
  - OpenCL: enabled by default only on Linux, because `mesa` is broken on Darwin;
  - CUDA: enabled when `cudaSupport = true;` is specified in the Nixpkgs config.

There is a weird input called "nvidiaComputeDrivers" that is necessary to test the OpenCL and CUDA backends on non-NixOS systems (because normally `nix build` can't see the system-installed `libcuda.so` or OpenCL vendors configs.

Part of #ZurichZHF :switzerland: 

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/arrayfire/default.nix [hunk 1]
```diff
@@ -1,91 +1,183 @@
...
+    repo = "arrayfire-data";
+    rev = "a5f533d7b864a4d8f0dd7c9aaad5ff06018c4867";
+    sha256 = "sha256-AWzhsrDXyZrQN2bd0Ng/XlE8v02x7QWTiFTyaAuRXSw=";
+  };
+  cub = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "cub";
+    rev = "1.10.0";
+    sha256 = "sha256-JyyNaTrtoSGiMP7tVUu9lFL07lyfJzRTVtx8yGy6/BI=";
+  };
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/arrayfire/default.nix:85]: This, iirc, is provided by `cudaPackages.cccl` (or something like that; also used by the nvidia-thrust derivation). Is it necessary to `cp -rf` the whole tree, or is it possible to discover it properly using e.g. CMake?

Comment (twesterhout) [reply pkgs/development/libraries/arrayfire/default.nix]: @SomeoneSerge good catch, thanks! Fixed. I've tested the change locally on a CUDA-capable GPU, and the tests still pass.

File: pkgs/development/libraries/arrayfire/default.nix [hunk 2]
```diff
@@ -1,91 +1,177 @@
...
-, cudaSupport ? config.cudaSupport
-, cudatoolkit
-, darwin
+, span-lite
+, stdenv
+, doCheck ? false
+, withCPU ? true
+, withCuda ? config.cudaSupport
+, withOpenCL ? stdenv.isLinux
+, nvidiaComputeDrivers ? null
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/arrayfire/default.nix:28]: How is this intended to be used?

File: pkgs/development/libraries/arrayfire/default.nix [hunk 3]
```diff
@@ -1,91 +1,177 @@
...
-, buildDocs ? false
-, config
-, cudaSupport ? config.cudaSupport
-, cudatoolkit
-, darwin
+, span-lite
+, stdenv
+, doCheck ? false
+, withCPU ? true
+, withCuda ? config.cudaSupport
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/arrayfire/default.nix:26]: I notice this changes the attribute name `cudaSupport -> withCuda` which people may have already been using. I see how the new naming is more consistent (`withCPU`, `withOpenCL`, so `withCuda`), but I suspect that `cudaSupport` is used more often across nixpkgs, and it's the status quo name since the expression has already been using it.

Personally, I don't mind renaming and I'll approve the PR even if you keep this extra change. I do think, however, there are more reasons to keep the old name than not to

File: pkgs/development/libraries/arrayfire/default.nix [hunk 4]
```diff
@@ -1,91 +1,177 @@
...
+    rev = "d2a66367d859cdb554f2374e29d39c88d5fff978";
+    hash = "sha256-9r1w0U9MvhduHwBpEWpqkrQPawd94EY3FAqSJghi09I=";
+  };
+
+  assets = fetchFromGitHub {
+    owner = pname;
+    repo = "assets";
+    rev = "cd08d749611b324012555ad6f23fd76c5465bd6c";
+    sha256 = "sha256-v4uhqPz1P1g1430FTmMp22xJS50bb5hZTeEX49GgMWg=";
+  };
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/arrayfire/default.nix:49]: A comment explaining what these are, "Assets for the ArrayFire documentation". I also wonder if there's a flag allowing to opt out of building the documentation (and thus out of carrying these PNGs around at runtime)

**Review by SomeoneSerge — COMMENTED** [review]
Thanks @twesterhout, this must've taken a lot of work to piece together! I'm asking a few more questions just to understand how the package is going to be updated in the future

File: pkgs/development/libraries/arrayfire/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/arrayfire/default.nix]: I don't see the "mark resolved" button, I guess you have it

File: pkgs/development/libraries/arrayfire/default.nix [hunk 5]
```diff
@@ -1,91 +1,177 @@
...
-    cudatoolkit
-  ] ++ lib.optionals buildDocs [
-    doxygen
-  ] ++ lib.optionals stdenv.isDarwin [
-    darwin.apple_sdk_11_0.frameworks.Accelerate
+    opencl-clhpp
+    span-lite
+  ]
+  ++ lib.optionals withCuda [
+    cudaPackages.cudatoolkit
```
[the remark this answers is not in the document]
Comment (twesterhout) [reply pkgs/development/libraries/arrayfire/default.nix:162]: I've opened an issue upstream: https://github.com/arrayfire/arrayfire/issues/3491

File: pkgs/development/libraries/arrayfire/default.nix (hunk 4 above)
Comment (twesterhout) [reply pkgs/development/libraries/arrayfire/default.nix]: I managed to get rid of these altogether.

File: pkgs/development/libraries/arrayfire/default.nix [hunk 6]
```diff
@@ -1,91 +1,203 @@
...
-, cudatoolkit
-, darwin
+  # OpenCL needs mesa which is broken on Darwin
+, openclSupport ? !stdenv.isDarwin
+  # This argument lets one run CUDA & OpenCL tests on non-NixOS systems by
+  # telling Nix where to find the drivers.
+, nvidiaComputeDrivers ? null
 }:
 
+assert blas.isILP64 == false;
```
Comment (chessai) [inline pkgs/development/libraries/arrayfire/default.nix:39]: can you add a comment here? why is this needed?

File: pkgs/development/libraries/arrayfire/default.nix [hunk 7]
```diff
@@ -1,91 +1,177 @@
...
+    repo = "CLBlast";
+    rev = "4500a03440e2cc54998c0edab366babf5e504d67";
+    sha256 = "sha256-I25ylQp6kHZx6Q7Ph5r3abWlQ6yeIHIDdS1eGCyArZ0=";
+  };
+  clfft = fetchFromGitHub {
+    owner = pname;
+    repo = "clfft";
+    rev = "760096b37dcc4f18ccd1aac53f3501a83b83449c";
+    sha256 = "sha256-vJo1YfC2AJIbbRj/zTfcOUmi0Oj9v64NfA9MfK8ecoY=";
+  };
```
[the remark this answers is not in the document]
Comment (chessai) [reply pkgs/development/libraries/arrayfire/default.nix:62]: Seconded

**Comment by twesterhout** [comment]
Hello @SomeoneSerge and @chessai, it's been a while (sorry), and I think it's time we finish and merge this. I've rebased the PR on the latest nixpkgs-unstable and ran all the tests (all three backends) on a local cluster node with NVidia V100 GPU (The system doesn't have NixOS, so using the GPU via Nix requires some hacks, and this is the reason the `nvidiaComputeDrivers` parameter exists -- see https://github.com/twesterhout/arrayfire-nix/blob/main/flake.nix#L21 and https://github.com/twesterhout/arrayfire-nix/blob/main/flake.nix#L83 for how it's used; sure, it's not ideal, but better than nothing I think...). I suggest we merge this and leave further improvements to other PRs.

**Comment by infinisil** [comment]
CI failure on Linux:
```
CMake Error at /build/source/build/extern/fg_glad-subbuild/fg_glad-populate-prefix/tmp/fg_glad-populate-gitupdate.cmake:34 (message):
  Failed to get the hash for HEAD:
  fatal: not a git repository: '.git'
make[2]: *** [CMakeFiles/fg_glad-populate.dir/build.make:126: fg_glad-populate-prefix/src/fg_glad-populate-stamp/fg_glad-populate-update_disconnected] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/fg_glad-populate.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
CMake Error at CMakeModules/ForgeFetchContent.cmake:798 (message):
  Build step for fg_glad failed: 2
Call Stack (most recent call first):
  CMakeModules/ForgeFetchContent.cmake:822 (__FetchContent_directPopulate)
  CMakeModules/ForgeConfigureDepsVars.cmake:104 (FetchContent_Populate)
  CMakeLists.txt:71 (fg_dep_check_and_populate)
-- Configuring incomplete, errors occurred!
error: builder for '/nix/store/vzv9vnzjzr5gvi9prk3da0spy87w8iy7-forge-1.0.8.drv' failed with exit code 1;
```

**Comment by infinisil** [comment]
@ofborg build arrayfire
